cds/container/hopscotch_hashmap.h

```diff
@@ -0,0 +1,660 @@
+				for (unsigned int iElm = 0; iElm < num_elm; ++iElm) {
+					if (Bucket::_empty_hop_info != (segments_arys + iElm)->_hop_info) {
+						counter += __popcnt((segments_arys + iElm)->_hop_info);
+					}
+				}
+				return counter;
+				*/
+			}
+
+			/// Checks whether the map contains \p key
+			/**
+			The function searches the item with key equal to \p key
+			and returns \p true if it is found, and \p false otherwise.
+			*/
+			template <typename K>
+			bool contains(K const& key)
+			{		
+				return find_with(key, [=](K const& one, K const& two) { return one != two; }, [](mapped_type&) {});
+			}
+
+			/// Checks whether the map contains \p key using \p pred predicate for searching
+			/**
+			The function is similar to <tt>contains( key )</tt> but \p pred is used for key comparing.
+			\p Less functor has the interface like \p std::less.
+			\p Less must imply the same element order as the comparator used for building the map.
+			*/
+			template <typename K, typename Predicate>
+			bool contains(K const& key, Predicate pred)
+			{
+				return find_with(key, pred, [](mapped_type&) {});
+			}
+
+			/// Find the key \p key
+			/** \anchor cds_nonintrusive_CuckooMap_find_func
+
+			The function searches the item with key equal to \p key and calls the functor \p f for item found.
+			The interface of \p Func functor is:
+			\code
+			struct functor {
+			void operator()( value_type& item );
+			};
+			\endcode
+			where \p item is the item found.
+
+			The functor may change \p item.second.
+
+			The function returns \p true if \p key is found, \p false otherwise.
+			*/
+			template <typename K, typename Func>
+			bool find_with(K const& key, Func f)
+			{
+				return find_with(key, [=](K const& one, K const& two) { return one != two; }, f);
+			}
+
+			template <typename K, typename Predicate>
+			bool find(K const& key, Predicate pred)
+			{
+				return find_with(key, pred, [](mapped_type&) {});
+			}
+
+			template <typename K>
+			bool find(K const& key)
+			{
+				return find_with(key, [=](K const& one, K const& two) { return one != two; }, [](mapped_type&) {});
+			}
+
+			template <typename K, typename Predicate, typename Func>
+			bool find_with(K const& key, Predicate pred, Func f)
+			{
+				unsigned int hash = calc_hash(key);
+				Bucket* start_bucket = segments_arys + hash;
+				unsigned int try_counter = 0;
+				unsigned int timestamp;
+				do {
+					timestamp = start_bucket->_timestamp;
+					unsigned int hop_info = start_bucket->_hop_info;
+					Bucket* check_bucket = start_bucket;
+					unsigned int temp;
+					for (int i = 0; i < HOP_RANGE; i++) {
+						temp = hop_info;
+						temp = temp >> i;
+
+						if (temp & 1) {
+							if (pred(key, *(check_bucket->_key)) == 0) {
+								f(*(check_bucket->_data));
+								return true;
+							}
+						}
+						++check_bucket;
+					}
+					++try_counter;
+				} while (timestamp != start_bucket->_timestamp && try_counter < MAX_TRIES);
```

Comment:
`MAX_TRIES` really shouldn't exist. It's a holdover from the initial version and it isn't a good addition. It breaks linearisability under heavy insertion and deletion contention. Keep trying until you get a timestamp match, there's no reason to have a limit on the times you check.

Reply:
@DaKellyFella, since hop_info is limited to the sizeof(unsigned int), we can't keep tring because we can get undefined behaviour. In case of offsets' implementation that could be true, but it seems that not in this implementation.

Reply:
@LeoSko 
> since hop_info is limited to the sizeof(unsigned int), we can't keep tring because we can get undefined behaviour.

No I mean `MAX_TRIES` shouldn't exist. You should keep rereading the bucket until you have completed a scan of the buckets `hop_info` indicates as useful until the timestamps haven't changed. That way you can guarantee nothing has changed since you scanned the table.

Does that make sense? Am I missing something? Apologies if I am.